internal/cosmetic/injector.go

```diff
@@ -0,0 +1,157 @@
+package cosmetic
+
+import (
+	"bytes"
+	"compress/flate"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"mime"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/andybalholm/brotli"
+	"github.com/anfragment/zen/internal/logger"
+	"github.com/klauspost/compress/zstd"
+	"golang.org/x/net/html/charset"
+)
+
+var (
+	// reBody captures contents of the body tag in an HTML document.
+	reBody          = regexp.MustCompile(`(?i)<body[\s\S]*?>([\s\S]*)</body>`)
+	styleOpeningTag = []byte("<style>")
+	styleClosingTag = []byte("</style>")
+)
+
+type Injector struct {
+	// store stores and retrieves css by hostname.
+	store Store
+}
+
+type Store interface {
+	Add(hostnames []string, selector string)
+	Get(hostname string) []string
+}
+
+func NewInjector(store Store) (*Injector, error) {
+	if store == nil {
+		return nil, errors.New("store is nil")
+	}
+
+	return &Injector{
+		store: store,
+	}, nil
+}
+
+func (inj *Injector) Inject(req *http.Request, res *http.Response) error {
+	hostname := req.URL.Hostname()
+	log.Printf("injecting css for %q\n", logger.Redacted(hostname))
+
+	selectors := inj.store.Get(hostname)
+	log.Printf("got %d selectors for %q", len(selectors), logger.Redacted(hostname))
+	if len(selectors) == 0 {
+		return nil
+	}
+
+	var ruleInjection bytes.Buffer
+	ruleInjection.Write(styleOpeningTag)
+	for _, selector := range selectors {
+		ruleInjection.WriteString(fmt.Sprintf("%s { display: none !important; }\n", selector))
+	}
+	ruleInjection.Write(styleClosingTag)
+
+	rawBodyBytes, err := readRawBody(res)
+	if err != nil {
+		return fmt.Errorf("read raw body: %w", err)
+	}
+
+	modifiedBody := reBody.ReplaceAllFunc(rawBodyBytes, func(body []byte) []byte {
+		// Inject ruleInjection right after the <body> tag.
+		return bytes.Join([][]byte{body, ruleInjection.Bytes()}, nil)
+	})
+
+	res.Body = io.NopCloser(bytes.NewReader(modifiedBody))
+	res.ContentLength = int64(len(modifiedBody))
+	res.Header.Set("Content-Length", fmt.Sprintf("%d", len(modifiedBody)))
```

Comment:
`strconv.Itoa` should be slightly faster since it doesn't need reflection to determine the type of the argument.